webgl.go

```
@@ -43,329 +44,353 @@ func DefaultAttributes() *ContextAttributes {
+	TRIANGLE_STRIP                               int
+	UNPACK_ALIGNMENT                             int
+	UNPACK_COLORSPACE_CONVERSION_WEBGL           int
+	UNPACK_FLIP_Y_WEBGL                          int
+	UNPACK_PREMULTIPLY_ALPHA_WEBGL               int
+	UNSIGNED_BYTE                                int
+	UNSIGNED_INT                                 int
+	UNSIGNED_SHORT                               int
+	UNSIGNED_SHORT_4_4_4_4                       int
+	UNSIGNED_SHORT_5_5_5_1                       int
+	UNSIGNED_SHORT_5_6_5                         int
+	VALIDATE_STATUS                              int
+	VENDOR                                       int
+	VERSION                                      int
+	VERTEX_ATTRIB_ARRAY_BUFFER_BINDING           int
+	VERTEX_ATTRIB_ARRAY_ENABLED                  int
+	VERTEX_ATTRIB_ARRAY_NORMALIZED               int
+	VERTEX_ATTRIB_ARRAY_POINTER                  int
+	VERTEX_ATTRIB_ARRAY_SIZE                     int
+	VERTEX_ATTRIB_ARRAY_STRIDE                   int
+	VERTEX_ATTRIB_ARRAY_TYPE                     int
+	VERTEX_SHADER                                int
+	VIEWPORT                                     int
+	ZERO                                         int
+}
+
+// initGlConstants initializes all GL constants from a webgl context.
+// This is necessary because wasm does not support struct tags.
+func (c *Context) initGlConstants() {
+
+	fields := reflect.TypeOf(*c)
+	num := fields.NumField()
+
+	// This functions is only called if creating a webgl context was successful.
+	// Thus we don't need to check here. Prototype is required for Safari.
+
+	ctx := js.Global().Get("WebGLRenderingContext").Get("prototype")
+
+	for i := 0; i < num; i++ {
+		field := fields.Field(i)
+		// Ignore the embedded js.Value. All other fields are gl constants.
+		if field.Name != "Value" {
+			// Retrieve value from gl context.
+			jsval := ctx.Get(field.Name)
+			if jsval.Type() == js.TypeNumber {
```

When is the type not number? If we cannot expect, should it return error or just panic?

---

E.g. the type is not a number if the GL constant does not exist in webgl. I discovered a few of them through this and removed them from the constants above.
I checked for "not number" so we can just ignore all missing values. In fact there should be no missing values now. Maybe we should just panic here if something is missing?

---

Sure, I prefer panic since the constants in the spec must be obtained.

It looks like the removed constants might be defined in extensions?

---

The constants I removed are the following:

INFO_LOG_LENGTH
not needed in webgl.

NUM_COMPRESSED_TEXTURE_FORMATS
maybe related to https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/Compressed_texture_formats

SHADER_COMPILER
no idea what this is

SHADER_SOURCE_LENGTH
not needed in webgl

STENCIL_INDEX
not in spec. see https://github.com/KhronosGroup/WebGL/issues/2370#issuecomment-294536185